rancher-cleanup/cleanup.sh

```diff
@@ -28,28 +35,55 @@ echo "=> Printing cluster info for confirmation"
 kubectl cluster-info
 kubectl get nodes -o wide
 
-echo -n "Do you want to continue (y/n)?"
-read answer
+if [ "$1" != "force" ]; then
+    echo -n "Do you want to continue (y/n)?"
+    read answer
 
-if [ "$answer" != "y" ]; then
-    exit 1
+    if [ "$answer" != "y" ]; then
+        exit 1
+    fi
 fi
 
 kcd()
 {
-  kubectl delete --ignore-not-found=true --grace-period=30 "$@"
+    i="0"
+    while [ $i -lt 4 ]; do
+        timeout 21 sh -c 'kubectl delete --ignore-not-found=true --grace-period=15 --timeout=20s '"$@"''
+        if [ $? -eq 0 ]; then
+            break
+        fi
+        i=$[$i+1]
+    done
 }
 
 kcpf()
 {
-  kubectl patch -p '{"metadata":{"finalizers":null}}' --type=merge "$@"
+  FINALIZERS=$(kubectl get -o jsonpath="{.metadata.finalizers}" "$@")
+  if [ "x${FINALIZERS}" != "x" ]; then
+      echo "Finalizers before for "$@": "${FINALIZERS}""
+      kubectl patch -p '{"metadata":{"finalizers":null}}' --type=merge "$@"
+      echo "Finalizers after for "$@": $(kubectl get -o jsonpath="{.metadata.finalizers}" "$@")"
+  fi
 }
 
 kcdns()
 {
   if kubectl get namespace $1; then
-    kubectl get namespace "$1" -o json | tr -d "\n" | sed "s/\"finalizers\": \[[^]]\+\]/\"finalizers\": []/"   | kubectl replace --raw /api/v1/namespaces/$1/finalize -f -
-    kubectl delete --ignore-not-found=true --grace-period=30 namespace $1
+    kcpf namespace $1
+    FINALIZERS=$(kubectl get -o jsonpath="{.spec.finalizers}" namespace "$1")
+    if [ "x${FINALIZERS}" != "x" ]; then
+        echo "Finalizers before for namespace "$1": "${FINALIZERS}""
+        kubectl get -o json namespace "$1" | tr -d "\n" | sed "s/\"finalizers\": \[[^]]\+\]/\"finalizers\": []/"   | kubectl replace --raw /api/v1/namespaces/$1/finalize -f -
+        echo "Finalizers after for namespace "$1": $(kubectl get -o jsonpath="{.spec.finalizers}" namespace "$1")"
```

Comment:
why do we need to patch the finalizer again after calling the `kcpf namespace $1` above?

Reply:
I've seen both exist on namespaces, `metadata.finalizers` and `spec.finalizers` (where `.spec.finalizers` can't be patched so it needs to be using this construction)